src/debug/telemetry/telemetry.c

```diff
@@ -21,6 +22,78 @@
 
 LOG_MODULE_DECLARE(ipc, CONFIG_SOF_LOG_LEVEL);
 
+struct perf_bitmap {
+	sys_bitarray_t *array;
+	uint16_t occupied;
+	size_t size;
+};
+
+static int perf_bitmap_init(struct perf_bitmap * const bitmap, sys_bitarray_t *array, size_t size)
+{
+	bitmap->array = array;
+	bitmap->size = size;
+	bitmap->occupied = 0;
+	return 0;
+}
+
+static int perf_bitmap_alloc(struct perf_bitmap * const bitmap, size_t *offset)
+{
+	int ret = sys_bitarray_alloc(bitmap->array, 1, offset);
+
+	if (!ret)
+		bitmap->occupied++;
```

Comment:
Ah, I see. This is used in comp_new and it seems that those can be executed on other cores so I should just add a spinlock for good measure. To be fair, as this code is based on Zephyr's bitarray, it would be a better idea to just add occupied bit count to bitarray in Zephyr, as it already has spinlocks and everything. Getting this accepted may take a while though. Might need to update this later.